packages/actions-shared/src/engage/utils/MessageSendPerformer.ts

```diff
@@ -155,66 +106,113 @@ export abstract class MessageSendPerformer<
-    let result: ModifiedResponse<unknown> | undefined = undefined
-    try {
-      this.statsIncr('perform')
-      result = await this.sendToRecepient(recepient)
-      return result
-    } catch (e) {
-      error = e
-    } finally {
-      try {
-        if (error && !isRetryableError(error)) {
-          const errorDetails = getErrorDetails(error)
-          if (errorDetails?.status) {
-            const cachedError = new CachedError(errorDetails.status, errorDetails.message, errorDetails.code)
-            await this.putCache(cachedError.serialize(), messageId, recipientId)
-          }
-        } else if (!error && result) {
-          const cachedResult = new CachedValue(result.status)
-          await this.putCache(cachedResult.serialize(), messageId, recipientId)
+      if (parsingError) {
+        //exception happened while parsing the cache.
+        // Log it and execute as if we don't have cache
+        this.logError('cache_parsing_error', { key, value: cacheReadValue, parsingError })
+        this.statsIncr('cache_parsing_error')
+      } else if (parsedCache) {
+        //parsed cache successfully && cache is not expired
+        // parsedValue - either value or error was parsed
+        if ('error' in parsedCache) {
+          this.statsIncr('cache_hit', 1, [`cached_error:${!!parsedCache?.error}`])
+          if (parsedCache?.error) throw parsedCache.error
+        }
+        if ('value' in parsedCache) {
+          return parsedCache.value
         }
-      } catch (cacheError) {
-        this.logError(`Failed to cache response for messageId: ${messageId}, recipientId: ${recipientId}`, cacheError)
+        throw new Error('Unexpected cache parsing result.')
+      } else {
+        //cache parsed successfully but cache needs to be ignored (e.g. expired) - re-execute
+        this.logError('cache_ignored', { key, value: cacheReadValue })
+        this.statsIncr('cache_ignored')
       }
     }
+    // re-executing, because cache not found or ignored or failed to read or parse
+    this.statsIncr('cache_miss')
+    this.logInfo('cache_miss', { key })
+    const result = await getOrCatch(() => createValue())
+
+    //before returning result - we need to try to serialize it and store it in cache
+    const stringified = getOrCatch(() =>
+      serializer.stringify('error' in result ? { error: result.error } : { value: result.value })
+    )
+    if ('error' in stringified && stringified.error) {
+      this.logError('cache_stringify_error', { key, error: stringified.error })
+      this.statsIncr('cache_stringify_error')
+    } else if ('value' in stringified && stringified.value) {
+      //result stringified and contains cacheable value - cache it
+      const result = getOrCatch(() => this.engageDestinationCache?.setByKey(key, stringified.value))
+      if ('error' in result && result.error) {
+        this.logError('cache_saving_error', { key, error: result.error })
+        this.statsIncr('cache_saving_error')
+      }
```

Comment:
bug in the code and notably happy typescript didn't help :-)
```suggestion
      const result = await getOrCatch(() => this.engageDestinationCache?.setByKey(key, stringified.value))
      if ('error' in result && result.error) {
        ...
      }
```
also var name result is confusable with line 193. And again, this is less readable code, comparing to what it was:
```
      const { error: cacheSavingError } = await getOrCatch(() => this.engageDestinationCache!.setByKey(key, stringified.value!)
      if (cacheSavingError) {
          ...
      }
      
```